### PR TITLE
Upgrade XGBoost to use BB on all platforms

### PR DESCRIPTION
**Pull request by staticfloat**
This upgrades the package (and a few stylistic points within the package itself) to use prebuilt binaries instead of requiring the user to compile on their own machine.

### CODE REVIEW

**Comment by staticfloat** [comment]
I've switched this to WIP, because in my testing, Windows doesn't work yet, and I also suspect that there are `std::string` values that cross the API boundary, which means we need to test out the new BB multiversioning capability to match against `std::string` ABI version.  Hold tight.  :)

**Comment by staticfloat** [comment]
So the problem with Windows is that the binaries immediately crash Julia upon `dlopen()`.  The backtrace is the following:

```
julia> using Libdl; Libdl.dlopen("xgboost.dll")
warning: Critical error detected c0000374

Thread 1 received signal SIGTRAP, Trace/breakpoint trap.
0x000007fba7a7a598 in ntdll!RtlpNtCreateKey () from /cygdrive/c/Windows/SYSTEM32/ntdll.dll
(gdb) bt
#0  0x000007fba7a7a598 in ntdll!RtlpNtCreateKey () from /cygdrive/c/Windows/SYSTEM32/ntdll.dll
#1  0x000007fba7a7f3f0 in ntdll!RtlpNtCreateKey () from /cygdrive/c/Windows/SYSTEM32/ntdll.dll
#2  0x000007fba7a25f05 in ntdll!RtlSetCurrentEnvironment () from /cygdrive/c/Windows/SYSTEM32/ntdll.dll
#3  0x000007fba54c11bc in msvcrt!free () from /cygdrive/c/Windows/system32/msvcrt.dll
#4  0x000000006fcb366d in libstdc++-6!_ZNSs6assignERKSs () from /home/Administrator/dist/bin/libstdc++-6.dll
#5  0x0000000029d96f9d in dmlc::data::LibSVMParserParam::__MANAGER__() () from /cygdrive/c/Users/Administrator/.julia/packages/XGBoost/VnexW/deps/usr/bin/xgboost.dll
#6  0x0000000029f0559e in _GLOBAL__sub_I_data.cc () from /cygdrive/c/Users/Administrator/.julia/packages/XGBoost/VnexW/deps/usr/bin/xgboost.dll
#7  0x0000000029da4ed5 in __do_global_ctors () at /workspace/srcdir/mingw-w64-v6.0.0/mingw-w64-crt/crt/gccmain.c:67
#8  0x0000000029d2129e in __DllMainCRTStartup (hDllHandle=0x29d20000, dwReason=1, lpreserved=0x0) at /workspace/srcdir/mingw-w64-v6.0.0/mingw-w64-crt/crt/crtdll.c:193
#9  0x000007fba79ab7df in ntdll!RtlAllocateAndInitializeSid () from /cygdrive/c/Windows/SYSTEM32/ntdll.dll
#10 0x000007fba79bd288 in ntdll!RtlLocaleNameToLcid () from /cygdrive/c/Windows/SYSTEM32/ntdll.dll
#11 0x000007fba79bc817 in ntdll!RtlLocaleNameToLcid () from /cygdrive/c/Windows/SYSTEM32/ntdll.dll
#12 0x000007fba79be5a6 in ntdll!RtlLocaleNameToLcid () from /cygdrive/c/Windows/SYSTEM32/ntdll.dll
#13 0x000007fba79be966 in ntdll!RtlLocaleNameToLcid () from /cygdrive/c/Windows/SYSTEM32/ntdll.dll
#14 0x000007fba79d1c47 in ntdll!LdrLoadDll () from /cygdrive/c/Windows/SYSTEM32/ntdll.dll
#15 0x000007fba4bc7c4d in LoadLibraryExW () from /cygdrive/c/Windows/system32/KERNELBASE.dll
#16 0x000000006b5af0ba in jl_dlopen (filename=filename@entry=0x29d1eaf0 "xgboost.dll", flags=flags@entry=68)
    at /home/Administrator/buildbot/worker/package_win64/build/src/dlload.c:85
#17 0x000000006b5af2b0 in jl_load_dynamic_library (modname=0xd675d58 "xgboost.dll", flags=68, throw_err=701623824)
    at /home/Administrator/buildbot/worker/package_win64/build/src/dlload.c:189
#18 0x00000000030e7a19 in julia_#dlopen#3_14362 () at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\Libdl\src\Libdl.jl:109
#19 0x0000000028d20e0a in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

**Comment by staticfloat** [comment]
Alright, I'm not certain why I thought `std::basic_string` values were somehow exempt from the `cxx11` string ABI problems, but they aren't.  Building string-abi-specific tarballs totally fixes this, and this package passes tests on linux64, mac64 and win64, so I am taking off the `WIP` tag.  :)

**Comment by aviks** [comment]
Thanks Elliot, this is exciting. Do you know if these binaries are built with openmp?

**Comment by staticfloat** [comment]
All except for MacOS are.  This is because on MacOS, we are using `clang` to build Julia, and getting OpenMP to work with `clang` is apparently complicated, I haven't looked into it too much.  Using `gcc` on MacOS will cause us to mix C++ stdlibs, and I don't think that's a good thing, so we're sticking with non-OpenMP on macOS.  Is that reasonable?

**Comment by aviks** [comment]
>Is that reasonable?

eminently.

File: Manifest.toml [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# This file is machine-generated - editing it directly is not advised
```
Comment (aviks) [inline Manifest.toml]: Any particular reason you want to check in the Manifest file? I thought best practice for projects was to only check in the Project file.

File: Project.toml [hunk 2]
```diff
@@ -3,14 +3,15 @@ uuid = "009559a3-9522-5dbb-924b-0b6ed2b22bb9"
...
+BinaryProvider = "b99e7846-7c00-51b0-8f62-c81ae34c0232"
+Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
 Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
 Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 
 [compat]
-BinDeps = "≥ 0.8.10"
```
Comment (aviks) [inline Project.toml]: should there be a BinaryProvider lower bound here?

Comment (staticfloat) [reply Project.toml]: We could put `0.3.0` as the lower bound.

File: src/XGBoost.jl [hunk 3]
```diff
@@ -9,7 +9,12 @@ export DMatrix, Booster
 export xgboost, predict, save, nfold_cv, slice, get_info, set_info, dump_model, importance
 export rabit_init, rabit_finalize, rabit_is_distributed, rabit_get_rank, rabit_get_world_size, rabit_get_version_number
 
+global const build_version = "0.82"
```
Comment (aviks) [inline src/XGBoost.jl]: can this be somehow generated in the build.jl, so that if we upgrade the dep, this upgrades as well? not a big deal.

Comment (staticfloat) [reply src/XGBoost.jl]: It's not generated right now, but you could put it in there if you want.

File: Manifest.toml (hunk 1 above)
Comment (staticfloat) [reply Manifest.toml]: Checking in the `Manifest.toml` gives insurance against someone who just wants to get things working in the future being surprised when there are no upper bounds and things don't work.  They can just `instantiate` and get the exact versions that a package had the last time someone dev'ed on it.  It's kind of like carrying around a "known good point", even if there are thousands of possible good points in the package space, here's at least one that works.

**Pull request merged**